Car/TeamNull SmartCar/src/util/HardwareUtil.cpp

```diff
@@ -61,12 +61,18 @@ void initialiseSensors() {
   delay(150);
   // Serial.println("00");
   frontSensor.setTimeout(500);
-  if (!frontSensor.init()) {
+  while (!frontSensor.init()) {
     Serial.println("Failed to initialise FRONT sensor");
-    blink(15);
-    while (1) {
-    }
+    blink(2);
+    pinMode(frontSensorPin, OUTPUT);
+    digitalWrite(frontSensorPin, LOW);
+
+    delay(1500);
+    digitalWrite(frontSensorPin, HIGH);
+    delay(200);
+    pinMode(frontSensorPin, INPUT);
```

Comment:
What does this do?

Reply:
Basically, when the car is powered by USB, everything boots correctly, however, when booting from batteries, the VL53L0Xs fail to initialise, this repeated powering them up and down was intended to see if they would eventually power-up successfully, after repeated attempts.
We believe therefore that the battery-based power supply has some sort of an issue where the ground is not at 0v, or something along those lines, although it is not easy to test this theory, as we do not have access to an oscilloscope, and the multimeter that we have does not seem to provide any value.

Configuring the pin as an output, with a low state will not damage the sensor, and, as the sensor has an internal pull-up resistor, configuring the pin as INPUT will cause it to be pulled up to 3.3v (to answer partially your next question)

The code configuring the pin high was never run on the car, as we had intended first to research if the board would handle this, before running it.

Reply:
>  however, when booting from batteries, the VL53L0Xs fail to initialise

You are the first group that comes up with this problem and generally considering the circumstances I find it highly unlikely there is an actual problem.
How are you connected to the sensors? Have you placed the sensors in the pre-determined sockets?

> Configuring the pin as an output, with a low state will not damage the sensor, and, as the sensor has an internal pull-up resistor, configuring the pin as INPUT will cause it to be pulled up to 3.3v (to answer partially your next question)

Yet again, how have you connected the sensors? Is the ESP32 board attached to the car? If so, then the pins that are exposed with the female sockets use 5V, NOT 3.3V. 

Please clarify or show your connections.

Reply:
(Please excuse any typo's, working on a laptop that doesn't have English installed)
- One VL53L0X works perfectly, as expected
- However, in order to use two (or more), as per the sensor's documentation, you have to cycle the XSHUT pins sequentially, in order to give each sensor a unique address. - I am not aware of any other groups that are currently using both provided VL53L0X sensors.

- The sensors are in the designated sockets
- However, according to the schematics of the circuit board that are provided, the XSHUT pins on the headers do not appear to be traced to any pins on the ESP32, and so we have connected them to the exposed female pin headers. As I said, the code as is, driving the pins high has not been run on the car, because of the 5V issue.
The pins are raised to 3.3V by configuring them as inputs, in combination with the sensors internal pull-up resistors.

- TL/DR: Sensor is in designated header, with additional wire from XSHUT to a female pin header port, not being driven high.

Not sure if this clarifies the connection, alternatively, I can get a picture of the connection.

Reply:
Oh yes it does and I just discovered that those sensors have a really **stupid** design. I was under the assumption that once you set an address the address persists which is the behavior of other i2c sensors. 😩😩
I guess I'm also stupid for not trying this out later but assuming it'd work. 😓😓😓
I mean, I selected those sensors literally because I thought I could put all of them on the i2c bus and waste no pins. Now it appears we'd have to waste one pin for each sensor if multiple were used.

In any case, unless you use the 3.3V pins on the board (which would be pretty hard considering they don't have female sockets) please don't put high voltage on the xshut pin unless you read somewhere it's OK to do so. If you need multiple sensors, use the ultrasonics.

Reply:
Glad we're on the same page now.
Yes, rather silly that the adresses are not persistent haha, not quite sure who came up with that, as it means you always need to waste a pin per sensor, defeating the i2c pin efficiency.

- Would you be willing to indicate on which pin of the level converters we may solder? Or alternatively, would you be willing to solder this for us, if we bring the car to you?

Please see the additional, seperate issue that I will log now before responding

Reply:
Let's not go into solutions that involve soldering.

> This pin is an active-low shutdown input; the board pulls it up to VDD to enable the sensor by default. Driving this pin low puts the sensor into hardware standby. This input is not level-shifted.
> [Source](https://www.pololu.com/product/2490)

Alright, so it looks like we can get away with the following:
* To put the sensor in standby mode the pin should be set as an `OUTPUT` and then `LOW` 
* To have the sensor working, the pin should be set as an `INPUT`

In other words, the pin is never set `HIGH` from the microcontroller's side and therefore no 5V ever reach the sensor.

Reply:
Yes, this is exactly how it was done, as I mentioned, the code as in this pull request was never loaded onto the car, as we were still investigating the possibility of safely using 5V.

When the car is powered by USB, the sensors initialise perfectly, first time, however, when powered by the batteries on the car, the sensors are unable to be initialised, which we suspect has some strange reference voltage issue.
The multimeter that we have, has however not provided useful insight into this, as it's own presence is enough to cause the sensors to initialise correctly, when trying to measure voltages.

Reply:
I'll try it out today or tomorrow on my setup and will let you know if I find something interesting.

Reply:
> 
> 
> I'll try it out today or tomorrow on my setup and will let you know if I find something interesting.

Thank you, looking forward to the result

Reply:
Just to verify, did you try the car with batteries on and a *single* sensor?

Reply:
Yes, on batteries, with a single sensor, the car is able to initialse correctly, my belief (unproven) is that the ESP32 pin set as output, low, is not going low enough to reset the sensor, when powered by batteries. If this is the case, it would hold that one sensor will initialise correctly, but two will not.

Reply:
Hi @platisd, any luck with your sensors? :)
TeamNull

Reply:
Didn't try yet :hear_no_evil:
I'll try this weekend.

Reply:
@platisd, I am marking this conversation as resolved, so that we can go ahead with the merge, in order to dissolve the DEV branch, but am aware that this issue is still being resolved, although this merge is not the issue as such.